File: Leetcode/1817. Finding the Users Active Minutes/solution.cpp
#include <vector>
#include <unordered_map>

class Solution {
public:
    vector<int> findingUsersActiveMinutes(vector<vector<int>>& logs, int k) {
        std::unordered_map<int, std::unordered_set<int>> userActions;

        for (const auto& log : logs)
        {
            int userId = log[0];
            int actionMin = log[1];

            userActions[userId].insert(actionMin);
        }

        std::vector<int> usersByUAM(k, 0);
        for (const auto& actions : userActions) 
        {
            int totalMins = actions.second.size();
            usersByUAM[totalMins - 1]++;
        }

        return usersByUAM;
    }
};

File: Leetcode/1817. Finding the Users Active Minutes/solution.py
class Solution:
    def findingUsersActiveMinutes(self, logs: List[List[int]], k: int) -> List[int]:
        users_by_uam = [0] * k
        user_actions = {}

        for log in logs:
            user_id, action_min = log
            if user_actions.get(user_id):
                user_actions[user_id].add(action_min)
            else:
                user_actions[user_id] = set([action_min])

        for _, actions in user_actions.items():
            total_mins = len(actions)
            users_by_uam[total_mins - 1] += 1
        
        return users_by_uam